cmd/benchfilesort/main.go

```diff
@@ -0,0 +1,454 @@
+	if err != nil {
+		return b, errors.Trace(err)
+	}
+
+	binary.BigEndian.PutUint64(head, uint64(len(body)))
+
+	b = append(b, head...)
+	b = append(b, body...)
+
+	return b, nil
+}
+
+func decodeRow(fd *os.File) (*comparableRow, error) {
+	var (
+		err  error
+		n    int
+		head = make([]byte, 8)
+		dcod = make([]types.Datum, 0, keySize+valSize+1)
+	)
+
+	n, err = fd.Read(head)
+	if n != 8 {
+		return nil, errors.New("incorrect header")
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	rowSize := int(binary.BigEndian.Uint64(head))
+	rowBytes := make([]byte, rowSize)
+
+	n, err = fd.Read(rowBytes)
+	if n != rowSize {
+		return nil, errors.New("incorrect row")
+	}
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	dcod, err = codec.Decode(rowBytes, keySize+valSize+1)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	return &comparableRow{
+		key:    dcod[:keySize],
+		val:    dcod[keySize : keySize+valSize],
+		handle: dcod[keySize+valSize:][0].GetInt64(),
+	}, nil
+}
+
+func encodeMeta(b []byte, scale int, keySize int, valSize int) []byte {
+	meta := make([]byte, 8)
+
+	binary.BigEndian.PutUint64(meta, uint64(scale))
+	b = append(b, meta...)
+	binary.BigEndian.PutUint64(meta, uint64(keySize))
+	b = append(b, meta...)
+	binary.BigEndian.PutUint64(meta, uint64(valSize))
+	b = append(b, meta...)
+
+	return b
+}
+
+func decodeMeta(fd *os.File) error {
+	meta := make([]byte, 24)
+	if n, err := fd.Read(meta); err != nil || n != 24 {
+		if n != 24 {
+			return errors.New("incorrect meta data")
+		}
+		return errors.Trace(err)
+	}
+
+	scale = int(binary.BigEndian.Uint64(meta[:8]))
+	if scale <= 0 {
+		return errors.New("number of rows must be positive")
+	}
+
+	keySize = int(binary.BigEndian.Uint64(meta[8:16]))
+	if keySize <= 0 {
+		return errors.New("key size must be positive")
+	}
+
+	valSize = int(binary.BigEndian.Uint64(meta[16:]))
+	if valSize <= 0 {
+		return errors.New("value size must be positive")
+	}
+
+	return nil
+}
+
+/*
+ * The synthetic data is exported as a binary format.
+ * The encoding format is:
+ *   1) Meta Data
+ *      Three 64-bit integers represent scale size, key size and value size.
+ *   2) Row Data
+ *      Each row is encoded as:
+ *		One 64-bit integer represent the row size in bytes, followed by the
+ *      the actual row bytes.
+ */
+func export() error {
+	var (
+		err         error
+		outputBytes []byte
+		outputFile  *os.File
+	)
+
+	fileName := path.Join(tmpDir, "data.out")
+	outputFile, err = os.OpenFile(fileName, os.O_WRONLY|os.O_CREATE|os.O_TRUNC, 0600)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	defer outputFile.Close()
+
+	outputBytes = encodeMeta(outputBytes, scale, keySize, valSize)
+
+	seed := rand.NewSource(time.Now().UnixNano())
+	r := rand.New(seed)
+
+	for i := 1; i <= scale; i++ {
+		outputBytes, err = encodeRow(outputBytes, nextRow(r, keySize, valSize))
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	_, err = outputFile.Write(outputBytes)
```

Comment:
Write for every row instead of once at the end, so we don't need to allocate large contiguous memory.